tokio/src/sync/semaphore.rs

```diff
@@ -1,1070 +1,183 @@
-
-        debug_assert_eq!(closed, prev.is_closed());
-
-        // The stub is only pushed when there are pending tasks. Because of
-        // this, the state must *always* be in pointer mode.
-        let prev = prev.waiter().unwrap();
-
-        // We don't want the *existing* pointer to be a stub.
-        debug_assert_ne!(prev, stub);
-
-        // Release `task` to the consume end.
-        prev.as_ref().next.store(stub.as_ptr(), Release);
-    }
-
-    fn stub(&self) -> NonNull<WaiterNode> {
-        unsafe { NonNull::new_unchecked(&*self.stub as *const _ as *mut _) }
-    }
-}
-
-impl fmt::Debug for Semaphore {
-    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
-        fmt.debug_struct("Semaphore")
-            .field("state", &SemState::load(&self.state, Relaxed))
-            .field("head", &self.head.with(|ptr| ptr))
-            .field("rx_lock", &self.rx_lock.load(Relaxed))
-            .field("stub", &self.stub)
-            .finish()
-    }
-}
-
-unsafe impl Send for Semaphore {}
-unsafe impl Sync for Semaphore {}
-
-// ===== impl Permit =====
-
-impl Permit {
-    /// Create a new `Permit`.
+    /// # Warning
     ///
-    /// The permit begins in the "unacquired" state.
-    pub(crate) fn new() -> Permit {
-        Permit {
-            waiter: None,
-            state: PermitState::Idle,
-        }
-    }
-
-    /// Returns true if the permit has been acquired
-    pub(crate) fn is_acquired(&self) -> bool {
-        self.state == PermitState::Acquired
-    }
-
-    /// Try to acquire the permit. If no permits are available, the current task
-    /// is notified once a new permit becomes available.
-    pub(crate) fn poll_acquire(
-        &mut self,
-        cx: &mut Context<'_>,
-        semaphore: &Semaphore,
-    ) -> Poll<Result<(), AcquireError>> {
-        match self.state {
-            PermitState::Idle => {}
-            PermitState::Waiting => {
-                let waiter = self.waiter.as_ref().unwrap();
-
-                if waiter.acquire(cx)? {
-                    self.state = PermitState::Acquired;
-                    return Ready(Ok(()));
-                } else {
-                    return Pending;
-                }
-            }
-            PermitState::Acquired => {
-                return Ready(Ok(()));
-            }
-        }
-
-        match semaphore.poll_permit(Some((cx, self)))? {
-            Ready(()) => {
-                self.state = PermitState::Acquired;
-                Ready(Ok(()))
-            }
-            Pending => {
-                self.state = PermitState::Waiting;
-                Pending
-            }
+    /// If the semaphore has less permits than the given `n` this function
+    /// will never return and wait for new permits to be added until `n`
+    /// permits were removed from the semaphore.
+    pub async fn remove_permits(&self, n: usize) -> Result<(), AcquireError> {
+        for _ in 0..n {
```

Comment:
Such an API is tricky, because it allows for deadlocks. Imagine a few tasks, which all try to acquire `n` permits. Since every task might get a few of them they will eventually all be stuck. A `Semaphore` should be fair towards users, recognize that a caller required a higher amount of permits and reserve them for the caller. In a similar fashion as a reader/writer lock must give the writer some priority in order to prevent it from ever getting scheduled.

I'm not sure if that is possible by building on the existing low level semaphore, or whether you would need a different implementation for it.

Reply:
It's also weird that this doesn't return a `Permit`, so the user manually needs to release everything.

Reply:
> Such an API is tricky, because it allows for deadlocks. Imagine a few tasks, which all try to acquire `n` permits.

Aquiring multiple permits at the same time is not currently in the scope of the Semaphore implementation. It would be a nice feature to have but also require a completely different implementation.

> It's also weird that this doesn't return a Permit, so the user manually needs to release everything.

`remove_permits` aquires `n` permits and calls `forget` on them. A forgotten permit is consumed and can't be returned. What do you mean by `the user manually needs to release everything`?

Reply:
I agree that w/ `Matthias` here, if we support `n` it should be an "atomic" op. Could we just leave this fn out for now and revisit later?

Reply:
Sure. I just wanted to have a counter part to `add_permits`. It had to be `async` because it needs to wait for all permits to become availabl, but there is nothing wrong with removing it or putting it behind a feature flag for the time being. It's simple to write this code in your own code if ever needed.